### PR TITLE
Port to Gtk4

### LINKED ISSUE #80: Port to Gtk4
### Objective:

Port Taxi to Gtk4 to take advantage of the improvements and new features this version offers.

### Advantages of Porting to Gtk4:

- Performance Improvements: Gtk4 provides enhanced performance compared to previous versions, resulting in a smoother and more responsive interface.

- New Functionalities: Gtk4 introduces new widgets and improvements to existing ones, allowing for a more modern and attractive user interface. For instance, the ability to use shadow effects and advanced animations.

- Wayland Compatibility: Transitioning to Gtk4 will facilitate native support for Wayland, improving compatibility with more modern desktop environments.

- Future Development: Updating to Gtk4 ensures the project's continuity and support, allowing Taxi to evolve alongside the latest Linux desktop technologies.

### Next Steps:

- Assess the compatibility of the existing code with Gtk4.
- Rewrite or adapt the user interface components to utilize the new Gtk4 APIs.
- Test and fix issues during the migration.
- Document the changes made and update dependencies.

### PR DESCRIPTION
**Pull request by alainm23**
Fix #80

- [x] Implement DnD
- [x] Refactor Gtk.Menu

### CODE REVIEW

File: src/Frontend/MainWindow.vala [hunk 1]
```diff
@@ -129,49 +136,47 @@ class Taxi.MainWindow : Hdy.ApplicationWindow {
...
         connect_box.connect_initiated.connect (on_connect_initiated);
         connect_box.ask_hostname.connect (on_ask_hostname);
         connect_box.bookmarked.connect (bookmark);
 
         file_operation.operation_added.connect (popover.add_operation);
         file_operation.operation_removed.connect (popover.remove_operation);
-        file_operation.ask_overwrite.connect (on_ask_overwrite);
-
-        key_press_event.connect (connect_box.on_key_press_event);
+        //  file_operation.ask_overwrite.connect (on_ask_overwrite);
```
Comment (danirabbit) [inline src/Frontend/MainWindow.vala:179]: oops, commented code

File: src/Frontend/MainWindow.vala [hunk 2]
```diff
@@ -71,50 +71,57 @@ class Taxi.MainWindow : Hdy.ApplicationWindow {
...
+        };
 
-        bookmark_list = new Gtk.Grid ();
-        bookmark_list.margin_top = bookmark_list.margin_bottom = 3;
-        bookmark_list.orientation = Gtk.Orientation.VERTICAL;
+        bookmark_list = new Gtk.Box (Gtk.Orientation.VERTICAL, 0) {
+            margin_top = 6,
+            margin_bottom = 6,
+            margin_start = 3,
+            margin_end = 3
```
Comment (danirabbit) [inline src/Frontend/MainWindow.vala:90]: Instead of doing this you can add `Granite.STYLE_CLASS_MENU` to the popover

File: src/Frontend/Widgets/FilePane.vala [hunk 3]
```diff
@@ -21,12 +21,12 @@ namespace Taxi {
...
     }
 
-    const Gtk.TargetEntry[] TARGET_LIST = {
-        { "test/plain", 0, Target.STRING },
-        { "text/uri-list", 0, Target.URI_LIST }
-    };
+    //  const Gtk.TargetEntry[] TARGET_LIST = {
+    //      { "test/plain", 0, Target.STRING },
+    //      { "text/uri-list", 0, Target.URI_LIST }
+    //  };
```
Comment (danirabbit) [inline src/Frontend/Widgets/FilePane.vala:27]: Oops more commented code

File: src/Frontend/Widgets/FilePane.vala [hunk 4]
```diff
@@ -42,48 +42,53 @@ namespace Taxi {
...
-            var placeholder_label = new Gtk.Label (_("This Folder Is Empty"));
-            placeholder_label.halign = Gtk.Align.CENTER;
-            placeholder_label.valign = Gtk.Align.CENTER;
-            placeholder_label.show ();
+            var placeholder_label = new Gtk.Label (_("This Folder Is Empty")) {
+                halign = Gtk.Align.CENTER,
+                valign = Gtk.Align.CENTER
+            };
+            placeholder_label.add_css_class (Granite.STYLE_CLASS_H2_LABEL);
+            placeholder_label.add_css_class (Granite.STYLE_CLASS_DIM_LABEL);
```
Comment (danirabbit) [inline src/Frontend/Widgets/FilePane.vala:54]: Can this just be a Granite.Placeholder?

**Review by danirabbit — CHANGES_REQUESTED** [review]
Left some comments but it looks like maybe this isn't actually ready for review? There's a lot of commented code

**Comment by alainm23** [comment]
I still have some features pending, but thanks for the comments.

**Comment by danirabbit** [comment]
Ah yeah it popped up in my notifications. Maybe better to set to draft until it's ready :)

File: src/meson.build [hunk 5]
```diff
@@ -18,15 +18,15 @@ executable(
     'Frontend/Widgets/ConnectBox.vala',
     'Frontend/Widgets/FilePane.vala',
     'Frontend/Widgets/OperationsPopover.vala',
-    'Frontend/Widgets/PathBarSeparator.vala',
     'Frontend/Widgets/PathBar.vala',
+    'Frontend/Widgets/FileRow.vala',
```
Comment (ryonakano) [inline src/meson.build:22]: Please update `po/POTFILES` too when you add or delete source files :)

File: src/meson.build [hunk 6]
```diff
@@ -18,15 +18,15 @@ executable(
...
     gresource,
     dependencies: [
         dependency('glib-2.0'),
         dependency('gobject-2.0'),
-        dependency('granite', version: '>=6.0.0'),
-        dependency('gtk+-3.0', version: '>=3.22'),
-        dependency('libhandy-1'),
+        dependency('granite-7'),
+        dependency('gtk4', version: '>=3.22'),
```
Comment (ryonakano) [inline src/meson.build:28]: You should simply remove `, version: '>=3.22'` or set correct required `4.*.*` version depending on the current code

**Comment by ryonakano** [comment]
Sorry, could you revert changes against the translation files themselves? These changes often make the PR bigger so we should update the translation files later in another PR or commit

File: src/Backend/FileOperations.vala [hunk 7]
```diff
@@ -149,4 +150,39 @@ namespace Taxi {
...
+
+        var replace_all_button = new Gtk.Button.with_label (_("Replace All Conflicts"));
+        message_dialog.add_action_widget (replace_all_button, ConflictFlag.REPLACE_ALL);
+
+        var skip_button = new Gtk.Button.with_label (_("Skip"));
+        message_dialog.add_action_widget (skip_button, ConflictFlag.SKIP);
+
+        var replace_button = new Gtk.Button.with_label (_("Replace"));
+        replace_button.add_css_class (Granite.STYLE_CLASS_SUGGESTED_ACTION);
+        message_dialog.add_action_widget (replace_button, ConflictFlag.REPLACE);
```
Comment (danirabbit) [inline src/Backend/FileOperations.vala:172]: `add_button` returns a Gtk.Widget so this is kind of an unnecessary refactor. For the one button that has a style class you can do:

```
        dialog.add_button (_("Replace All Conflicts"), ConflictFlag.REPLACE_ALL);
        dialog.add_button (_("Skip"), ConflictFlag.SKIP);
        var replace_button = dialog.add_button (_("Replace"), ConflictFlag.REPLACE);
        replace_button.add_css_class (Granite.STYLE_CLASS_SUGGESTED_ACTION);
```

File: src/meson.build (hunk 6 above)
Comment (ryonakano) [reply src/meson.build:28]: @alainm23 Is there any reason you didn't fix this?

File: src/Frontend/Widgets/FilePane.vala [hunk 8]
```diff
@@ -113,32 +116,44 @@ namespace Taxi {
...
-            foreach (Gtk.Widget row in list_box.get_children ()) {
-                if (row.get_data<Gtk.CheckButton> ("checkbutton").get_active ()) {
-                    uri_list.add (current_uri.to_string () + "/" + row.get_data<string> ("name"));
+
+            Gtk.ListBoxRow row = null;
+            var row_index = 0;
+
+            do {
+                row = list_box.get_row_at_index (row_index);
+                if (((FileRow)row).active) {
```
Comment (ryonakano) [inline src/Frontend/Widgets/FilePane.vala:125]: Here causes the following warning on runtime:

```
** (com.github.alecaddd.taxi:10910): CRITICAL **: 23:39:56.053: taxi_file_row_get_active: assertion 'self != NULL' failed
```

This is because we call `get_row_at_index` with off-by-one index after last row and there are null check here

Comment (ryonakano) [reply src/Frontend/Widgets/FilePane.vala:125]: You should add the type check or null check and plus avoid do-while like this:

```vala
            unowned var row = list_box.get_first_child ();
            while (row != null) {
                if (row is FileRow) {
                    unowned var file_row = (FileRow) row;
                    if (file_row.active) {
                        uri_list.add (current_uri.to_string () + "/" + file_row.file_name);
                    }
                }

                row = row.get_next_sibling ();
            }
```

File: src/Frontend/Widgets/FilePane.vala [hunk 9]
```diff
@@ -113,32 +116,44 @@ namespace Taxi {
...
+
             return uri_list;
         }
 
         public void update_list (GLib.List<FileInfo> file_list) {
             clear_children (list_box);
             // Have to convert to gee list because glib list sort function is buggy
             // (it randomly removes items...)
-            var gee_list = glib_to_gee<FileInfo> (file_list);
+            var gee_list = glib_to_gee<GLib.FileInfo> (file_list);
```
Comment (ryonakano) [inline src/Frontend/Widgets/FilePane.vala:139]: Is this change necessary for porting?

File: src/Frontend/Widgets/FilePane.vala [hunk 10]
```diff
@@ -113,32 +116,44 @@ namespace Taxi {
...
                     continue;
                 }
 
-                var row = new_row (file_info);
+                var row = new FileRow (file_info);
+                row.current_uri = current_uri;
+                row.on_checkbutton_toggle.connect (on_checkbutton_toggle);
+                row.on_delete.connect (() => {
+
+                });
```
Comment (ryonakano) [inline src/Frontend/Widgets/FilePane.vala:151]: What is this for?

File: src/Frontend/MainWindow.vala [hunk 11]
```diff
@@ -221,25 +234,30 @@ class Taxi.MainWindow : Hdy.ApplicationWindow {
...
-
+                var bookmark_item = new Gtk.Button () {
+                    child = new Gtk.Label (uri) {
+                        halign = START
+                    }
+                };
+                bookmark_item.add_css_class (Granite.STYLE_CLASS_MENUITEM);
                 bookmark_item.clicked.connect (() => {
                     connect_box.go_to_uri (uri);
+                    bookmark_popover.popdown ();
```
Comment (ryonakano) [inline src/Frontend/MainWindow.vala:256]: Doesn't `Gtk.Popover.autohide` work for you instead of changing the scope of `bookmark_popover` from a local variable to a private member variable and manually popdown here?

File: src/Frontend/MainWindow.vala [hunk 12]
```diff
@@ -69,52 +70,66 @@ class Taxi.MainWindow : Hdy.ApplicationWindow {
...
+            pixel_size = 64
+        };
+
+        title_label = new Gtk.Label (_("Connect"));
+        title_label.add_css_class (Granite.STYLE_CLASS_H1_LABEL);
+
+        var description_label = new Gtk.Label (_("Type a URL and press 'Enter' to\nconnect to a server.")) {
+            justify = CENTER
+        };
+
```
Comment (ryonakano) [inline src/Frontend/MainWindow.vala:129]: You can remove these custom widgets thanks to Granite.Placeholder

File: src/Frontend/MainWindow.vala [hunk 13]
```diff
@@ -69,52 +70,66 @@ class Taxi.MainWindow : Hdy.ApplicationWindow {
...
+        welcome_box = new Granite.Placeholder (_("Connect")) {
```
Comment (ryonakano) [inline src/Frontend/MainWindow.vala:130]: I prefer to name this variable as `welcome` to match `Granite.Widgets.Welcome` in the previous code to (possibly) reduce the diff and plus `Granite.Placeholder` is not a Box
